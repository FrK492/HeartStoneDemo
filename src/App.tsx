import React, { FC } from 'react';
import { NavigationContainer } from '@react-navigation/native';
import { createNativeStackNavigator } from '@react-navigation/native-stack';
import ListScreen from './screens/ListScreen';
import DetailScreen from './screens/DetailScreen'
import { store } from './redux/store'
import { Provider } from 'react-redux'

const Stack = createNativeStackNavigator();

const App: FC = () => {
  return (
    <Provider store={store}>
      <NavigationContainer>
        <Stack.Navigator>
          <Stack.Screen
            options={{title: 'Mechanics List'}}
            name='list_screen'
            component={ListScreen}
          />
          <Stack.Screen
            options={({ route }: any) => ({ title: route.params.mechanicName })}
            name='detail_screen'
            component={DetailScreen}
          />
        </Stack.Navigator>
      </NavigationContainer>
    </Provider>
  )
}

export default App