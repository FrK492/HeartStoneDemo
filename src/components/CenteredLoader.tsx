import React, { FC } from 'react'
import { View, ActivityIndicator, StyleSheet } from 'react-native'

interface IProps {
    size: 'small' | 'large'
}

export const CenteredLoader: FC<IProps> = (props) => {
    return (
        <View style={styles.container}>
            <ActivityIndicator size={props.size} />
        </View>
    )
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center'
    }
})