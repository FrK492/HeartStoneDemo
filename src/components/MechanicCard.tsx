import React, { FC } from 'react'
import { TouchableOpacity, Text, StyleSheet } from 'react-native'

interface IProps {
    mechanicName: string
    onCardPress: (mechanicName: string) => void
}

export const MechanicCard: FC<IProps> = (props) => {
    return (
        <TouchableOpacity onPress={() => props.onCardPress(props.mechanicName)} style={styles.container}>
            <Text>{props.mechanicName}</Text>
        </TouchableOpacity>
    )
}

const styles = StyleSheet.create({
    container: {
        margin: 10,
        padding: 10,
        backgroundColor: 'white',
        borderRadius: 5
    }
})