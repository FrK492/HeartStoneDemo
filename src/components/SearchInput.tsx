import React, { FC } from 'react'
import { View, StyleSheet, TextInput } from 'react-native'

interface IProps {
    value: string,
    onValueChange: (text: string) => void
    onSubmit: (text: string) => void
}

export const SearchInput: FC<IProps> = (props) => {
    return (
        <View style={styles.container}>
            <TextInput
                value={props.value}
                style={styles.input}
                placeholder='Search Cards'
                onChangeText={(text) => props.onValueChange(text)}
                onSubmitEditing={(event) => props.onSubmit(event.nativeEvent.text)}
                returnKeyLabel='Search'
                returnKeyType='search'
            />
        </View>
    )
}

const styles = StyleSheet.create({
    container: {
        height: 50,
        justifyContent: 'center',
        backgroundColor: 'white',
        margin: 5,
        borderRadius: 5
    },
    input: {
        paddingHorizontal: 10
    }
})