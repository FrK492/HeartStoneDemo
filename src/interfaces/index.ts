export interface IHeartStoneResponse {
    [key: string]: IHeartStoneCard[];
}

export interface IHeartStoneCard {
    cardId: string;
    dbfId: string;
    cardSet: string;
    type: string;
    locale: string;
    faction: string;
    name: string;
    playerClass: string;
    text: string;
    img?: string;
    cost?: number;
    attack?: number;
    health?: number;
    artist?: string;
    mechanics?: IMechanics[];
}

export interface IMechanics {
    name: string;
}