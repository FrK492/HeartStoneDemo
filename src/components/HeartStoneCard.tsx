import React, { FC } from 'react'
import { View, Text, Image, StyleSheet, FlatList } from 'react-native'
import { IHeartStoneCard } from '../interfaces'

interface IProps {
    card: IHeartStoneCard
}

export const HeartStoneCard: FC<IProps> = (props) => {
    const { card } = props
    return (
        <View style={styles.container}>
            <View style={styles.infoColumn}>
                {card.name && <Text><Text style={styles.infoTitle}>Card Name: </Text>{card.name}</Text>}
                {card.faction && <Text><Text style={styles.infoTitle}>Card Faction: </Text>{card.faction}</Text>}
                {card.cardSet && <Text><Text style={styles.infoTitle}>Card Set: </Text>{card.cardSet}</Text>}
                {card.text && <Text><Text style={styles.infoTitle}>Card Text: </Text>{card.text}</Text>}
                {card.type && <Text><Text style={styles.infoTitle}>Card Type: </Text>{card.type}</Text>}
                {card.mechanics &&
                    <>
                        <Text style={styles.infoTitle}>Mechanics: </Text>
                        <FlatList
                            data={card.mechanics}
                            keyExtractor={(item) => `${card.cardId}-${item.name}`}
                            renderItem={({item}) => <Text>{item.name}</Text>}
                        />
                    </> 
                }
            </View>
            <View style={styles.imageColumn}>
                {card.img && <Image source={{uri: card.img}} style={styles.cardImage} />}
            </View>
        </View>
    )
}

const styles = StyleSheet.create({
    container: {
        margin: 10,
        padding: 10,
        backgroundColor: 'white',
        borderRadius: 5,
        minHeight: 200,
        flexDirection: 'row',
        justifyContent: 'space-between'
    },
    infoColumn: {
        flex: 2
    },
    imageColumn: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center'
    },
    cardImage: {
        height: 100,
        width: 100,
        resizeMode: 'contain'
    },
    infoTitle: {
        fontSize: 14,
        fontWeight: 'bold',
        color: 'black'
    }
})