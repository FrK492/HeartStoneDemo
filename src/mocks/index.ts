import { IHeartStoneCard } from "../interfaces"

export const mockCard: IHeartStoneCard = {
    cardId: '123',
    cardSet: 'testSet',
    dbfId: '123',
    faction: 'testFaction',
    locale: 'tr',
    name: 'testName',
    playerClass: 'testClass',
    text: 'testText',
    type: 'testType'
}

export const mockCard2: IHeartStoneCard = {
    cardId: '1234',
    cardSet: 'testSet2',
    dbfId: '1234',
    faction: 'testFaction3',
    locale: 'en',
    name: 'testName2',
    playerClass: 'testClass2',
    text: 'testText2',
    type: 'testType2',
    img: 'https://dummyimage.com/300',
    mechanics: [
        { name: 'testMechanic' }
    ]
}

export const mockApiResponse = {
    testProp: [mockCard],
    testProp2: [mockCard2]
}

export const mockCardArray = [
    mockCard, mockCard2
]
