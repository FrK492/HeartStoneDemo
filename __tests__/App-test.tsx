import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import { mockApiResponse, mockCard, mockCard2, mockCardArray } from '../src/mocks';

// Components
import { CenteredLoader } from '../src/components/CenteredLoader';
import { SearchInput } from '../src/components/SearchInput';
import { MechanicCard } from '../src/components/MechanicCard';
import { HeartStoneCard } from '../src/components/HeartStoneCard';

// Utils
import {
  convertHeartStoneResponseToArray,
  exportMechanicsFromHeartStoneCards,
  filterHeartStoneCardsByMechanic
} from '../src/utils/arrayManipluation';


describe('Components', () => {
  it('Loader renders corretly', () => {
    const jsonTree = renderer.create(<CenteredLoader size='large' />).toJSON()
    expect(jsonTree).toMatchSnapshot()
  });

  it('SearchInput renders correctly', () => {
    const jsonTree = renderer.create(<SearchInput value='' onSubmit={() => {}} onValueChange={() => {}} />).toJSON()
    expect(jsonTree).toMatchSnapshot()
  })

  it('MechacnicCard renders correctly', () => {
    const jsonTree = renderer.create(<MechanicCard mechanicName='' onCardPress={() => {}} />).toJSON()
    expect(jsonTree).toMatchSnapshot()
  })

  it('HeartStoneCard renders correctly', () => {
    const jsonTree = renderer.create(<HeartStoneCard card={mockCard2} />).toJSON()
    expect(jsonTree).toMatchSnapshot()
  })
})

describe('Utils', () => {
  it('filterHeartStoneCardsByMechanic functions correctly', () => {
    const result = filterHeartStoneCardsByMechanic('testMechanic', mockCardArray)
    expect(result[0]).toBe(mockCard2)
  })

  it('exportMechanicsFromHeartStoneCards functions correctly', () => {
    const result = exportMechanicsFromHeartStoneCards(mockCardArray)
    expect(result[0]).toBe(mockCard2.mechanics!![0].name)
  })

  it('convertHeartStoneResponseToArray functions correctly', () => {
    const result = convertHeartStoneResponseToArray(mockApiResponse)
    expect(result[0]).toBe(mockCard)
  })

})

