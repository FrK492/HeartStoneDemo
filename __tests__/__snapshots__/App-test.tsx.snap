// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components HeartStoneCard renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderRadius": 5,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "margin": 10,
      "minHeight": 200,
      "padding": 10,
    }
  }
>
  <View
    style={
      Object {
        "flex": 2,
      }
    }
  >
    <Text>
      <Text
        style={
          Object {
            "color": "black",
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
      >
        Card Name: 
      </Text>
      testName2
    </Text>
    <Text>
      <Text
        style={
          Object {
            "color": "black",
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
      >
        Card Faction: 
      </Text>
      testFaction3
    </Text>
    <Text>
      <Text
        style={
          Object {
            "color": "black",
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
      >
        Card Set: 
      </Text>
      testSet2
    </Text>
    <Text>
      <Text
        style={
          Object {
            "color": "black",
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
      >
        Card Text: 
      </Text>
      testText2
    </Text>
    <Text>
      <Text
        style={
          Object {
            "color": "black",
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
      >
        Card Type: 
      </Text>
      testType2
    </Text>
    <Text
      style={
        Object {
          "color": "black",
          "fontSize": 14,
          "fontWeight": "bold",
        }
      }
    >
      Mechanics: 
    </Text>
    <RCTScrollView
      data={
        Array [
          Object {
            "name": "testMechanic",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      viewabilityConfigCallbackPairs={Array []}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <Text>
            testMechanic
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "https://dummyimage.com/300",
        }
      }
      style={
        Object {
          "height": 100,
          "resizeMode": "contain",
          "width": 100,
        }
      }
    />
  </View>
</View>
`;

exports[`Components Loader renders corretly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    size="large"
  />
</View>
`;

exports[`Components MechacnicCard renders correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "white",
      "borderRadius": 5,
      "margin": 10,
      "opacity": 1,
      "padding": 10,
    }
  }
>
  <Text />
</View>
`;

exports[`Components SearchInput renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderRadius": 5,
      "height": 50,
      "justifyContent": "center",
      "margin": 5,
    }
  }
>
  <TextInput
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Search Cards"
    returnKeyLabel="Search"
    returnKeyType="search"
    style={
      Object {
        "paddingHorizontal": 10,
      }
    }
    value=""
  />
</View>
`;
